modules/service/src/main/scala/lucuma/odb/service/CalibrationsService.scala

```diff
@@ -74,17 +94,134 @@ object CalibrationsService {
+                        mode = PosAngleConstraintMode.AverageParallactic.some, none
+                      ).some,
+                      observingMode = ObservingModeInput.Create(none, conf.some).some,
+                      scienceRequirements =
+                        ScienceRequirementsInput(
+                          mode = ScienceMode.Spectroscopy.some,
+                          spectroscopy = SpectroscopyScienceRequirementsInput.Default.copy(
+                            signalToNoise = Nullable.NonNull(SignalToNoise.unsafeFromBigDecimalExact(100.0))
+                        ).some
+                      ).some
+                    ).some
+            )
+          )
+        }
+
+        (for {
+          currGmosNLS <- session.execute(Statements.selectGmosNorthLongSlitConfigurations(true))(pid)
+          currGmosSLS <- session.execute(Statements.selectGmosSouthLongSlitConfigurations(true))(pid)
+          o1          <- gmosNorthLSObservations(gnls.diff(currGmosNLS))
+          o2          <- gmosSouthLSObservations(gsls.diff(currGmosSLS))
+        } yield (o1 ::: o2)).map(_.sequence)
+      }
+
+      // Set tthe calibration role of the observations in bulk
+      private def setCalibRoleAndGroup(oids: List[Observation.Id], calibrationRole: CalibrationRole)(using Transaction[F]): F[Unit] = {
+        val update = void"UPDATE t_observation " |+|
+          sql"SET c_calibration_role = $calibration_role "(calibrationRole) |+|
+          void"WHERE c_observation_id IN (" |+|
+            oids.map(sql"$observation_id").intercalate(void", ") |+| void")"
+        session.executeCommand(update).void
+      }
+
+      private def generateCalibrations(pid: Program.Id, gnls: List[GmosNConfigs], gsls: List[GmosSConfigs])(using Transaction[F]): F[Unit] = {
+        for {
+          cg  <- calibrationsGroup(pid, gnls.size + gsls.size)
+          _   <- cg.map(g =>
+                   calibrationObservations(pid, g, gnls, gsls).flatMap {
+                     case Result.Success(ids) if ids.nonEmpty =>
+                       setCalibRoleAndGroup(ids, CalibrationRole.SpectroPhotometric).void
+                     case _                                   =>
+                       Applicative[F].unit
+                   }
+                 ).getOrElse(Applicative[F].unit)
+        } yield ()
+      }
+
       def recalculateCalibrations(pid: Program.Id)(using Transaction[F]): F[Unit] =
         for {
-          gnls <- session.execute(Statements.SelectGmosNorthLongSlitConfigurations)(pid)
-          gsls <- session.execute(Statements.SelectGmosSouthLongSlitConfigurations)(pid)
-          _    <-  calibrationsGroup(pid, gnls.size + gsls.size)
+          gnls <- session.execute(Statements.selectGmosNorthLongSlitConfigurations(false))(pid)
+          gsls <- session.execute(Statements.selectGmosSouthLongSlitConfigurations(false))(pid)
+          _    <- generateCalibrations(pid, gnls, gsls).whenA(gnls.nonEmpty || gsls.nonEmpty)
```

Comment:
What happens when there are already calibrations in the group?  I didn't see the code to remove existing ones.

Reply:
at the moment it only creates if there is no exisiting one. I have a task also to prune them if needed